tarantool/crud.py

```diff
@@ -0,0 +1,748 @@
+class Response(object):
+    def __init__(self, response):
+        """
+        TODO: write docs
+        """
+
+        if isinstance(response, dict):
+            for response_field_name in response.keys():
+                setattr(self, response_field_name, response[response_field_name])
+        elif isinstance(response, str):
+            self.str = response
+            self.err = response
+
+
+class Result(Response):
+    """
+    Contains result's fields from result variable of crud module operation.
+    """
+
+
+class Error(Response):
+    """
+    Contains error's fields from error variable of crud module operation.
+    """
+
+
+class Crud(object):
+    """
+    Contains crud module methods.
+    """
+
+    def __init__(self, con):
+        """
+        TODO: write docs
+        """
+
+        self.con = con
+        self._crud_module_init_complete = False
```

Comment:
As far as I understand, there are two main reasons for `Crud` class:
- to store `_crud_module_init_complete` variable,
- to be able to use module as a namespace of `conn`: `conn.crud`, since you'll need conn anyway.

I think it is possible to get rid of both of them.

Above I've discussed that there are seems to be no need to call `crud = require('crud')`. You may check that the instance is a crud router (we may call any operations on routers only, we have no interest in interacting with storages explicitly) by `rawget(_G, 'crud') != nil`. But do we really need that? It seems like an optimization, but it is rather hard to have our "cache" consistent: if one would disable crud router (highly unlikely, but not impossible), it will show that require is ok which is untrue. And the main point is we actually don't need to know if we're able to call crud procedures: if we're don't, we will get "Procedure not defined" (code 33) error. It is rather simple to make this error more useful: just wrap each code 33 error with comment "Ensure that you're calling crud.router and user have enough grants". It seems no use to fast fail since if user would try to call `crud` thousands of times from a connection to an instance without `crud` they are doing something really weird and it's not worth to optimize it.

Reply:
Done, integrated functions directly into Connection (conn.crud_*). Also, when an `DatabaseError` exception with code 33 appears, I supplement it with auxiliary information

Reply:
Related issues:

* https://github.com/tarantool/crud/issues/229
* https://github.com/tarantool/cartridge-java/issues/159